src/libopenrave/kinbodygeometry.cpp

```diff
@@ -165,26 +162,54 @@ void AppendBoxTriangulation(const Vector& pos, const Vector& ex, TriMesh& tri)
     tri.indices.insert(tri.indices.end(), &indices[0], &indices[nindices]);
 }
 
-void AppendCylinderTriangulation(const Vector& pos, const dReal& rad, const dReal& len, const int& numverts, TriMesh& tri) {
+static void AppendConicalFrustumTriangulation(const Vector& pos, const dReal topRad, const dReal botRad, const dReal halfHeight, const uint numFaces, TriMesh& tri)
```

Comment:
still slower than when using index
```
rdiankov@PF3VMBC7:~ $ ./a.out 
using indices: 66060[us]
using std::vector::iterator: 80034[us]
rdiankov@PF3VMBC7:~ $ ./a.out 
using indices: 68879[us]
using std::vector::iterator: 84180[us]
rdiankov@PF3VMBC7:~ $ ./a.out 
using indices: 65620[us]
using std::vector::iterator: 79354[us]
rdiankov@PF3VMBC7:~ $ ./a.out 
using indices: 65254[us]
using std::vector::iterator: 78898[us]
rdiankov@PF3VMBC7:~ $ ./a.out 
using indices: 65312[us]
using std::vector::iterator: 79268[us]
rdiankov@PF3VMBC7:~ $ ./a.out 
using indices: 69626[us]
using std::vector::iterator: 82956[us]
rdiankov@PF3VMBC7:~ $ ./a.out 
using indices: 65528[us]
using std::vector::iterator: 78486[us]
```

Reply:
Looks like the problem with the original algorithm's duplicate vertices is that it started at `i=0`. If you set `i=1`, then there's no duplication:
```c++
void AppendCylinderTriangulationIndex(const Vector& pos, const dReal topRadius, const dReal bottomRadius, const dReal len, const int numverts, TriMesh& tri)
{
    // once again, cylinder is on z axis
    dReal dtheta = 2 * PI / (dReal)numverts;
    int base = tri.vertices.size();
    tri.vertices.resize(2*(numverts+1) + 4);
    tri.indices.resize((numverts+1) * 12);

    int vindex = 0;
    int iindex = 0;
    tri.vertices[vindex++] = Vector(0,0,len) + pos;
    tri.vertices[vindex++] = Vector(0,0,-len) + pos;
    tri.vertices[vindex++] = Vector(topRadius,0,len) + pos;
    tri.vertices[vindex++] = Vector(bottomRadius,0,-len) + pos;
    for(int i = 1; i < numverts+1; ++i) {
        dReal s = sinf(dtheta * (dReal)i);
        dReal c = cosf(dtheta * (dReal)i);
        int off = (int)tri.vertices.size();
        tri.vertices[vindex++] = Vector(topRadius*c, topRadius*s, len) + pos;
        tri.vertices[vindex++] = Vector(bottomRadius*c, bottomRadius*s, -len) + pos;

        tri.indices[iindex++] = (base);       tri.indices[iindex++] = (off-2);       tri.indices[iindex++] = (off);
        tri.indices[iindex++] = (base+1);       tri.indices[iindex++] = (off+1);       tri.indices[iindex++] = (off-1);
        tri.indices[iindex++] = (off-2);   tri.indices[iindex++] = (off-1);       tri.indices[iindex++] = (off);
        tri.indices[iindex++] = (off);     tri.indices[iindex++] = (off-1);       tri.indices[iindex++] = (off+1);
    }
}
```

Reply:
fixed indexing issue